crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -3986,65 +3957,56 @@ impl AuthorityPerEpochStore {
         last: Option<CheckpointHeight>,
     ) -> SuiResult<Vec<(CheckpointHeight, PendingCheckpointV2)>> {
         let tables = self.tables()?;
-        let mut iter = tables.pending_checkpoints_v2.unbounded_iter();
+        let mut db_iter = tables.pending_checkpoints_v2.unbounded_iter();
         if let Some(last_processed_height) = last {
-            iter = iter.skip_to(&(last_processed_height + 1))?;
+            db_iter = db_iter.skip_to(&(last_processed_height + 1))?;
         }
-        Ok(iter.collect())
+
+        let db_results: Vec<_> = db_iter.collect();
+
+        let mut quarantine_results = self
+            .consensus_quarantine
+            .read()
+            .get_pending_checkpoints(last);
+
+        // retain only the checkpoints with heights greater than the highest height in the db
+        if let Some(db_highest_height) = db_results.last().map(|(h, _)| h) {
+            quarantine_results.retain(|(h, _)| h > db_highest_height);
+        }
+
+        let mut db_results = db_results;
+        db_results.extend(quarantine_results);
+        Ok(db_results)
     }
 
-    pub fn get_pending_checkpoint(
-        &self,
-        index: &CheckpointHeight,
-    ) -> SuiResult<Option<PendingCheckpointV2>> {
-        Ok(self.tables()?.pending_checkpoints_v2.get(index)?)
+    pub fn pending_checkpoint_exists(&self, index: &CheckpointHeight) -> SuiResult<bool> {
+        Ok(self
+            .consensus_quarantine
+            .read()
+            .pending_checkpoint_exists(index))
     }
 
     pub fn process_pending_checkpoint(
```

Comment:
I know this is not relevant to this PR, but I wonder why we call this function process_*pending*_checkpoint, while in fact it process constructed checkpoint.

Reply:
good point, will rename